From 90dc141559b69f5ac23c3604297889423e998ae2 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 22 Nov 2012 21:30:59 -0500
Subject: [PATCH] Misc EMC Changes

Only spawn 1 ocelot ever from eggs.
Load custom permissions before plugins
Unload plugins after everything else
Make Durability enchant scale 0-80% reduction on Armor damage (up to level 5)
Zombies break down doors on normal
Update inventory on block place/interact cancellation
Don't update server.properties
Remove spigot update check
Supress empty itemstack log messages
Bat's with Passengers won't sleep
Disable thorns for arrows
---
 src/main/java/net/minecraft/server/EntityArrow.java   |  2 +-
 src/main/java/net/minecraft/server/EntityBat.java     |  2 +-
 src/main/java/net/minecraft/server/EntityOcelot.java  |  1 +
 src/main/java/net/minecraft/server/ItemStack.java     | 11 ++++++++++-
 .../net/minecraft/server/PathfinderGoalBreakDoor.java |  2 +-
 .../java/net/minecraft/server/PropertyManager.java    |  2 +-
 .../bukkit/craftbukkit/event/CraftEventFactory.java   |  1 +
 7 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/src/main/java/net/minecraft/server/EntityArrow.java b/src/main/java/net/minecraft/server/EntityArrow.java
index 9c97edf9c9..f2db28e4b7 100644
--- a/src/main/java/net/minecraft/server/EntityArrow.java
+++ b/src/main/java/net/minecraft/server/EntityArrow.java
@@ -396,7 +396,7 @@ public abstract class EntityArrow extends Entity implements IProjectile {
                 }
 
                 if (!this.world.isClientSide && entity1 instanceof EntityLiving) {
-                    EnchantmentManager.a(entityliving, entity1);
+                    //EnchantmentManager.a(entityliving, entity1); // EMC
                     EnchantmentManager.b((EntityLiving) entity1, (Entity) entityliving);
                 }
 
diff --git a/src/main/java/net/minecraft/server/EntityBat.java b/src/main/java/net/minecraft/server/EntityBat.java
index 34239160be..ef16b9657f 100644
--- a/src/main/java/net/minecraft/server/EntityBat.java
+++ b/src/main/java/net/minecraft/server/EntityBat.java
@@ -144,7 +144,7 @@ public class EntityBat extends EntityAmbient {
             this.yaw += f1;
             if (this.random.nextInt(100) == 0 && this.world.getType(blockposition1).isOccluding(this.world, blockposition1)) {
                 // CraftBukkit Start - Call BatToggleSleepEvent
-                if (CraftEventFactory.handleBatToggleSleepEvent(this, false)) {
+                if (!this.isVehicle() && CraftEventFactory.handleBatToggleSleepEvent(this, false)) { // EMC - add !isVehicle
                     this.setAsleep(true);
                 }
                 // CraftBukkit End
diff --git a/src/main/java/net/minecraft/server/EntityOcelot.java b/src/main/java/net/minecraft/server/EntityOcelot.java
index d9a7b8ac1e..673b655324 100644
--- a/src/main/java/net/minecraft/server/EntityOcelot.java
+++ b/src/main/java/net/minecraft/server/EntityOcelot.java
@@ -10,6 +10,7 @@ public class EntityOcelot extends EntityAnimal {
     private static final DataWatcherObject<Boolean> bx = DataWatcher.a(EntityOcelot.class, DataWatcherRegistry.i);
     private EntityOcelot.a<EntityHuman> by;
     private EntityOcelot.b bz;
+    public boolean spawnBonus = false; // Spigot // EMC
 
     public EntityOcelot(EntityTypes<? extends EntityOcelot> entitytypes, World world) {
         super(entitytypes, world);
diff --git a/src/main/java/net/minecraft/server/ItemStack.java b/src/main/java/net/minecraft/server/ItemStack.java
index 182633bca5..7d21657f2a 100644
--- a/src/main/java/net/minecraft/server/ItemStack.java
+++ b/src/main/java/net/minecraft/server/ItemStack.java
@@ -395,7 +395,16 @@ public final class ItemStack {
                 j = EnchantmentManager.getEnchantmentLevel(Enchantments.DURABILITY, this);
                 int k = 0;
 
+                j = Math.min(5, j); // EMC
                 for (int l = 0; j > 0 && l < i; ++l) {
+                    // EMC start - make Unbreaking scale to level for armor. 5 = 80% chance of no loss.
+                    if (getItem() instanceof ItemArmor) {
+                        if (random.nextFloat() >= 1 - (.8 * (j / 5))) {
+                            k++;
+                        }
+                        continue;
+                    }
+                    // EMC end
                     if (EnchantmentDurability.a(this, j, random)) {
                         ++k;
                     }
@@ -408,7 +417,7 @@ public final class ItemStack {
                     event.getPlayer().getServer().getPluginManager().callEvent(event);
 
                     if (i != event.getDamage() || event.isCancelled()) {
-                        event.getPlayer().updateInventory();
+                        //event.getPlayer().updateInventory(); // EMC - Handle in Event
                     }
                     if (event.isCancelled()) {
                         return false;
diff --git a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
index 7c7eadbe2c..d4e4e25582 100644
--- a/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
+++ b/src/main/java/net/minecraft/server/PathfinderGoalBreakDoor.java
@@ -65,7 +65,7 @@ public class PathfinderGoalBreakDoor extends PathfinderGoalDoorInteract {
             this.b = i;
         }
 
-        if (this.a == this.f() && this.a(this.entity.world.getDifficulty())) {
+        if (this.a == this.f() && this.entity.world.getDifficulty().ordinal() >= EnumDifficulty.NORMAL.ordinal()) { // EMC
             // CraftBukkit start
             if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityBreakDoorEvent(this.entity, this.door).isCancelled()) {
                 this.c();
diff --git a/src/main/java/net/minecraft/server/PropertyManager.java b/src/main/java/net/minecraft/server/PropertyManager.java
index 729455ce53..d12e2757e7 100644
--- a/src/main/java/net/minecraft/server/PropertyManager.java
+++ b/src/main/java/net/minecraft/server/PropertyManager.java
@@ -75,7 +75,7 @@ public abstract class PropertyManager<T extends PropertyManager<T>> {
     public void savePropertiesFile(java.nio.file.Path java_nio_file_path) {
         try {
             // CraftBukkit start - Don't attempt writing to file if it's read only
-            if (java_nio_file_path.toFile().exists() && !java_nio_file_path.toFile().canWrite()) {
+            if (java_nio_file_path.toFile().exists()) { // EMC - Never overwrite the file
                 return;
             }
             // CraftBukkit end
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index c0fd2948e3..ebc640957d 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -456,6 +456,7 @@ public class CraftEventFactory {
         }
         craftServer.getPluginManager().callEvent(event);
 
+        if (action == Action.RIGHT_CLICK_BLOCK && !cancelledBlock && event.isCancelled()) { player.updateInventory(); } // EMC
         return event;
     }
 
-- 
2.24.1

